sympy/matrices/expressions/tests/test_matrix_exprs.py

```diff
@@ -242,3 +243,8 @@ def test_Zero_power():
     assert z2**3 == MatPow(z2, 3).doit()
     assert z2**0 == Identity(3)
     raises(ValueError, lambda:MatPow(z2, -1).doit())
+
+
+def test_issue_7421():
+    assert isinstance(diff((D*w)[k,0], w[p,0]), Derivative)
+    assert isinstance(diff(w.T*D*w, w[0,0]), Derivative)
```

Comment:
I see, it's just the first one that uses Dummy, and it comes from the MatrixElement returning a Sum. So this is unrelated to this PR. To fix this (which would be a separate PR), I would suggest allowing MatrixElement to be unevaluated by default.